src/ray/stats/metric.h

```diff
@@ -191,6 +201,152 @@ struct MetricPoint {
+                      const std::vector<double> &buckets) {}
+
+template <StatsType T, StatsType... Ts>
+void RegisterViewList(const std::string &name, const std::string &description,
+                      const std::vector<opencensus::tags::TagKey> &tag_keys,
+                      const std::vector<double> &buckets) {
+  RegisterView<T>(name, description, tag_keys, buckets);
+  RegisterViewList<Ts...>(name, description, tag_keys, buckets);
+}
+
+inline std::vector<opencensus::tags::TagKey> convertTags(
+    const std::vector<std::string> &names) {
+  std::vector<opencensus::tags::TagKey> ret;
+  for (auto &n : names) {
+    ret.push_back(TagKeyType::Register(n));
+  }
+  return ret;
+}
+
+class Stats {
+  using Measure = opencensus::stats::Measure<double>;
+
+ public:
+  Stats(const std::string &measure, const std::string &description,
+        std::vector<std::string> tag_keys, std::vector<double> buckets,
+        std::function<void(const std::string &, const std::string,
+                           const std::vector<opencensus::tags::TagKey>,
+                           const std::vector<double> &buckets)>
+            register_func)
+      : tag_keys_(tag_keys) {
+    auto f = [register_func, measure, description, buckets, this]() {
+      measure_ = std::make_unique<Measure>(Measure::Register(measure, description, ""));
+      register_func(measure, description, convertTags(tag_keys_), buckets);
+    };
+
+    if (StatsConfig::instance().IsInitialized()) {
+      f();
+    } else {
+      StatsConfig::instance().AddInitializer(f);
+    }
+  }
+
+  void Record(double val) { Record(val, std::unordered_map<std::string, std::string>()); }
+
+  void Record(double val, std::string tag_val) {
+    RAY_CHECK(tag_keys_.size() == 1);
+    std::unordered_map<std::string, std::string> tags{{tag_keys_[0], std::move(tag_val)}};
+    Record(val, std::move(tags));
+  }
+
+  void Record(double val, std::unordered_map<std::string, std::string> tags) {
+    if (StatsConfig::instance().IsStatsDisabled() || !measure_) {
+      return;
+    }
+    TagsType combined_tags = StatsConfig::instance().GetGlobalTags();
+    // In case that tag containing non-printable chars
+    for (auto &t : tags) {
+      for (auto &c : t.second) {
+        if (!isprint(c)) {
+          c = '?';
+        }
+      }
+      combined_tags.emplace_back(TagKeyType::Register(t.first), std::move(t.second));
+    }
+    opencensus::stats::Record({{*measure_, val}}, combined_tags);
+  }
+
+ private:
+  std::unique_ptr<opencensus::stats::Measure<double>> measure_;
+  std::vector<std::string> tag_keys_;
+};
+
+}  // namespace details
+
 }  // namespace stats
 
 }  // namespace ray
+
+#define STATS_DEPAREN(X) STATS_ESC(STATS_ISH X)
+#define STATS_ISH(...) ISH __VA_ARGS__
+#define STATS_ESC(...) STATS_ESC_(__VA_ARGS__)
+#define STATS_ESC_(...) STATS_VAN##__VA_ARGS__
+#define STATS_VANISH
+
+#define DEFINE_stats(name, description, tags, buckets, types...)                     \
```

Comment:
Maybe calling it `REGISTER_stats`? DEFINE & DECLARE sounds a bit similar to me. It is your call.

Reply:
I did it on purpose to make it sounds like DEFINE_flags/DECLARE_flags so the end user knows what it is.